cpp/include/raft/neighbors/detail/cagra/add_nodes.cuh

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include <cstdint>
+#include <raft/core/device_resources.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/neighbors/cagra_types.hpp>
+#include <rmm/device_buffer.hpp>
+
+// prototype declaration
+namespace raft::neighbors::cagra {
+template <typename T, typename IdxT>
+void search(raft::resources const& res,
+            const raft::neighbors::cagra::search_params& params,
+            const raft::neighbors::cagra::index<T, IdxT>& idx,
+            raft::device_matrix_view<const T, int64_t, raft::row_major> queries,
+            raft::device_matrix_view<IdxT, int64_t, raft::row_major> neighbors,
+            raft::device_matrix_view<float, int64_t, raft::row_major> distances);
+}
+
+namespace raft::neighbors::cagra::detail {
+template <class CounterT, class IdxT>
+__global__ void count_incoming_edges_kernel(CounterT* const count_ptr,
+                                            const IdxT* const graph_ptr,
+                                            const std::size_t num_entries)
+{
+  for (std::size_t i = threadIdx.x + blockDim.x * blockIdx.x; i < num_entries;
+       i += gridDim.x * blockDim.x) {
+    const auto node_id = graph_ptr[i];
+    atomicAdd(count_ptr + node_id, 1);
+  }
+}
+
+template <class CounterT, class IdxT>
+void count_incoming_edges(CounterT* const count_ptr,
```

Comment:
Can we just use [histogram](https://github.com/rapidsai/raft/blob/9fb05a2ab3d72760a09f1b7051e711d773682ef1/cpp/include/raft/stats/histogram.cuh#L62)?

Reply:
Updated to use the histogram function.